src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+
+    /// Convenience class that overloads similar collection functionality
+    class BsonContainer : public std::variant<BsonDocument, BsonArray> {
+    public:
+        using base = std::variant<BsonDocument, BsonArray>;
+        using base::base;
+        using base::operator=;
+
+        void push_back(std::pair<std::string, Bson> value) {
+            if (std::holds_alternative<BsonDocument>(*this)) {
+                std::get<BsonDocument>(*this)[value.first] = value.second;
+            } else {
+                std::get<BsonArray>(*this).push_back(value.second);
+            }
+        }
+
+        std::pair<std::string, Bson> back()
+        {
+            if (std::holds_alternative<BsonDocument>(*this)) {
+                auto pair = *std::get<BsonDocument>(*this).back();
+
+                return pair;
+            } else {
+                return {"", std::get<BsonArray>(*this).back()};
+            }
+        }
+
+        void pop_back()
+        {
+            if (std::holds_alternative<BsonDocument>(*this)) {
+                std::get<BsonDocument>(*this).pop_back();
+            } else {
+                std::get<BsonArray>(*this).pop_back();
+            }
+        }
+
+        size_t size()
+        {
+            if (std::holds_alternative<BsonDocument>(*this)) {
+                return std::get<BsonDocument>(*this).size();
+            } else {
+                return std::get<BsonArray>(*this).size();
+            }
+        }
+    };
+
+    std::stack<BsonContainer> m_marks;
+    std::stack<Instruction> m_instructions;
+public:
+    using number_integer_t = typename json::number_integer_t;
+    using number_unsigned_t = typename json::number_unsigned_t;
+    using number_float_t = typename json::number_float_t;
+    using string_t = typename json::string_t;
+
+    parser() {
+        // use a vector container to hold any fragmented values
+        m_marks.push(BsonArray());
+    }
+
+    /*!
+    @brief a null value was read
+    @return whether parsing should proceed
+    */
+    bool null() override {
+        auto instruction = m_instructions.top();
+        m_instructions.pop();
+        check_state(instruction.type, state_t::json_key);
+        m_marks.top().push_back({instruction.key, bson_null});
+        return true;
+    }
+
+    /*!
+    @brief a boolean value was read
+    @param[in] val  boolean value
+    @return whether parsing should proceed
+    */
+    bool boolean(bool val) override {
+        auto instruction = m_instructions.top();
+        m_instructions.pop();
+        check_state(instruction.type, state_t::json_key);
+        m_marks.top().push_back({instruction.key, val});
+        return true;
+    }
+
+    /*!
+    @brief an integer number was read
+    @param[in] val  integer value
+    @return whether parsing should proceed
+    */
+    bool number_integer(number_integer_t) override {
+        throw BsonError(util::format("Invalid SAX instruction for canonical extended json: 'number_integer'"));
```

Comment:
So far as I can see, there is a design conflict for error checking, none of these methods ever return false, instead if there's an error they throw an exception. Unless I'm missing something, we should choose either a status checking API or an exception throwing one, but not both.

Reply:
I've effectively chosen an exception throwing API that uses status checks. I've no choice but to adhere to nlohman's SAX parser, but you'll notice that I always return true.